runtime/compiler/optimizer/TreeLowering.cpp

```diff
@@ -1355,43 +1386,70 @@ StoreArrayElementTransformer::lower(TR::Node* const node, TR::TreeTop* const tt)
    TR::SymbolReference *elementSymRef = comp->getSymRefTab()->findOrCreateArrayShadowSymbolRef(TR::Address, anchoredArrayBaseAddressNode);
    TR::Node *elementStoreNode = TR::Node::createWithSymRef(TR::awrtbari, 3, 3, elementAddress, anchoredValueNode, anchoredArrayBaseAddressNode, elementSymRef);
 
-   TR::SymbolReference *arrayStoreCHKSymRef = comp->getSymRefTab()->findOrCreateTypeCheckArrayStoreSymbolRef(comp->getMethodSymbol());
-   TR::Node *arrayStoreCHKNode = TR::Node::createWithRoomForThree(TR::ArrayStoreCHK, elementStoreNode, 0, arrayStoreCHKSymRef);
-   arrayStoreCHKNode->copyByteCodeInfo(node);
-
-   TR::TreeTop *arrayStoreCHKTT = originalBlock->append(TR::TreeTop::create(comp, arrayStoreCHKNode));
-
-   if (enableTrace)
-      traceMsg(comp, "Created arrayStoreCHK treetop n%dn node n%dn\n", arrayStoreCHKTT->getNode()->getGlobalIndex(), arrayStoreCHKNode->getGlobalIndex());
+   TR::Node *arrayStoreCHKNode = NULL;
+   TR::TreeTop *arrayStoreTT = NULL;
+   bool addArrayStoreCheck = !skipArrayStoreChecks(comp, node);
 
+   // Some JCL methods are known never to trigger an ArrayStoreException
+   // Only add an ArrayStoreCHK if it's required
+   if (addArrayStoreCheck)
+      {
+      TR::SymbolReference *arrayStoreCHKSymRef = comp->getSymRefTab()->findOrCreateTypeCheckArrayStoreSymbolRef(comp->getMethodSymbol());
+      arrayStoreCHKNode = TR::Node::createWithRoomForThree(TR::ArrayStoreCHK, elementStoreNode, 0, arrayStoreCHKSymRef);
+      arrayStoreCHKNode->copyByteCodeInfo(node);
+      arrayStoreTT = originalBlock->append(TR::TreeTop::create(comp, arrayStoreCHKNode));
+      if (enableTrace)
+         traceMsg(comp, "Created arrayStoreCHK treetop n%dn arrayStoreCHKNode n%dn\n", arrayStoreTT->getNode()->getGlobalIndex(), arrayStoreCHKNode->getGlobalIndex());
+      }
+   else
+      {
+      arrayStoreTT = originalBlock->append(TR::TreeTop::create(comp, TR::Node::create(TR::treetop, 1, elementStoreNode)));
```

Comment:
I remember I had to add `incReferenceCount` to `elementStoreNode` when ArrayStoreCheck is skipped, otherwise there is a crash. Without `incReferenceCount` when ArrayStoreCheck is skipped, one issue I found is that the reference count for `awrtbari` is not set up correctly. I wonder why is `incReferenceCount` removed here? 

```
   if (addArrayStoreCheck)
      {
...
      arrayStoreCHKTT = originalBlock->append(TR::TreeTop::create(comp, arrayStoreCHKNode));
      }
   else
      {
      arrayStoreCHKTT = originalBlock->append(TR::TreeTop::create(comp, elementStoreNode));
      elementStoreNode->incReferenceCount();
      }
```

Reply:
@a7ehuo I am confused by this last snippet of code. Nodes that appear at the top of the tree (as seems to be the case with `elementStoreNode`) should have a reference count of 0. So it seems odd to have the `incReferenceCount` call. Note that the code in this commit from Henry is creating a `treetop` opcode and that is obviously a difference between this commit and the code that you pasted above.

Reply:
>Note that the code in this commit from Henry is creating a treetop opcode and that is obviously a difference between this commit and the code that you pasted above

I think this might be the reason that in Henry's version, there is no need to add `elementStoreNode->incReferenceCount()`. `TR::Node::create(TR::treetop, 1, elementStoreNode)` will increment the reference count on `elementStoreNode`. `TR::TreeTop::create(comp, elementStoreNode)` doesn't. @hzongaro Could you confirm? 

A bit more on the original issue that I ran into with [the snippet of the code on the above](https://github.com/eclipse-openj9/openj9/pull/13299#discussion_r685450580): originally I found if ArrayStoreCheck is skipped, the `rc` for`awrtbari` is wrong as below: rc=1. It's supposed to be rc=2 since it's referenced again by compressedRefs `n1160n`.  If ArrayStoreCheck is not skipped, the `rc` for`awrtbari` is incremented by `createWithRoomForThree` when we create `arrayStoreCHKNode`. That's why at the time of the initial implementation, `incReferenceCount` is added when ArrayStoreCheck is skipped. 

```
n1154n    awrtbari  <array-shadow>[#229  Shadow] [flags 0x80000607 0x0 ] (sharedMemory )      [0x7ff633355870] bci=[2,37,-] rc=1 vc=0 vn=- li=- udi=- nc=3 flg=0x20
n1153n      aladd                                                                             [0x7ff633355820] bci=[2,37,-] rc=1 vc=0 vn=- li=- udi=- nc=2
n892n         ==>aRegLoad
n1152n        ladd (X>=0 )                                                                    [0x7ff6333557d0] bci=[0,12,-] rc=1 vc=0 vn=- li=- udi=- nc=2 flg=0x100
n1150n          lshl                                                                          [0x7ff633355730] bci=[0,12,-] rc=1 vc=0 vn=- li=- udi=- nc=2
n1148n            i2l                                                                         [0x7ff633355690] bci=[0,12,-] rc=1 vc=0 vn=- li=- udi=- nc=1
n895n               ==>iRegLoad
n1149n            iconst 2                                                                    [0x7ff6333556e0] bci=[-1,0,-] rc=1 vc=0 vn=- li=- udi=- nc=0
n1151n          lconst 16                                                                     [0x7ff633355780] bci=[-1,0,-] rc=1 vc=0 vn=- li=- udi=- nc=0
n893n       ==>aRegLoad
n892n       ==>aRegLoad
n1160n    compressedRefs                                                                      [0x7ff633355a50] bci=[2,37,-] rc=0 vc=0 vn=- li=- udi=- nc=2
n1154n      ==>awrtbari
n1159n      lconst 0       
```

Reply:
As Annabelle described, in earlier versions of my changes for VP and her changes for Tree Lowering, if the `ArrayStoreCHK` was not needed, the `awrtbari` was being placed at the root of a tree.  We both ran into intermittent failures with that arrangement of the IL, and explicitly incremented the reference count to resolve the problem.

A couple of weeks ago, I noticed a message in a log file regarding the root of a tree (an `awrtbari`) having a non-zero reference count.  That's when I realized that placing the `awrtbari` under a `treetop` node if the `ArrayStoreCHK` was not needed might be more correct than incrementing the node's reference count.  I made that change in both VP and Tree Lowering.

Sorry for not alerting you to that change, Annabelle @a7ehuo!

Reply:
Thank you for the clarification @hzongaro! The original crash makes more sense now. 

I'm curious of what the rules are on the reference count for a node. I noticed: 
- Treetop nodes or nodes at the root of a tree generally start with `rc=0` if it's not referenced anywhere else. 
- Non-tree top nodes start with `rc=1`  if it's not referenced anywhere else. 
- The exceptions are: when `awrtbari` is the treetop node or at the root of a tree, it starts with` rc=1`. Why does `awrtbari` have `rc=1` when it's a treetop node or at the root of a tree? Any other exceptions? 

Is my observation correct?

Reply:
I don't know of other exceptions to the treetop nodes have reference count of zero rule. I wish we did not have any exceptions to that rule but it may not be high enough priority to clean up right away. Could be a task for the backlog though. 

Going back to @a7ehuo IL trees shown, I still don't get why the reference count on the `awrtbari` has to be 2 in the case that you showed. The reference count for a given node is a simple concept : how many parent nodes does that node have ? In the IL trees that were shown above, there is only one parent for the `awrtbari` node and that is the `compressedRefs` node. So I don't understand why reference count 1 did not work. When you had an array store check node in addition to the `compressedRefs` node, then there were two parents for the `awrtbari` node and so a reference count of 2 made sense in that case.

Reply:
@vijay I see, just double checking my understanding: The rule is that treetop nodes should have reference count of zero because reference count shows how many parent nodes that a node has and treetop nodes do not have any parent node. Is that correct? Thanks!

Reply:
@a7ehuo that is correct in general, except for the `awrtbari` exception (a treetop node that also has a parent). Of course, the reference count should be equal to the number of parents in ALL cases.